pkg/docker/docker.go

```diff
@@ -0,0 +1,277 @@
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) EnvDir() string {
+	return filepath.Join(l.RootDir, "env")
+}
+
+func (l *Launchable) EnvVars() map[string]string {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) Executables(_ *runit.ServiceBuilder) ([]launch.Executable, error) {
+	// We don't make use of runit for docker launchables, so there are no
+	// executables to supply. Instead we'll just send "docker run" and
+	// "docker stop" commands to dockerd.
+	return nil, nil
+}
+
+func (l *Launchable) GetRestartTimeout() time.Duration {
+	return l.RestartTimeout
+}
+
+func (l *Launchable) ID() launch.LaunchableID {
+	return l.LaunchableID
+}
+
+func (l *Launchable) InstallDir() string {
+	// There is no install directory for a docker container
+	return ""
+}
+
+func (l *Launchable) Installed() bool {
+	// docker pull is idempotent and cheap, so don't worry about this
+	return false
+}
+
+func (l *Launchable) Launch(_ *runit.ServiceBuilder, _ runit.SV) error {
+	if l.DockerClient == nil {
+		return util.Errorf("docker client was not initialized, can't launch docker launchable")
+	}
+
+	ctx := context.TODO()
+	envVars := []string{}
+	envFiles, err := ioutil.ReadDir(l.EnvDir())
+	if err != nil {
+		return util.Errorf("could not read from %s: %v", l.EnvDir(), err)
+	}
+	for _, file := range envFiles {
+		if !file.IsDir() {
+			// read content and load into envVars
+			envFilePath := filepath.Join(l.EnvDir(), file.Name())
+			data, err := ioutil.ReadFile(envFilePath)
+			if err != nil {
+				return util.Errorf("could not read file %s: %v", envFilePath, err)
+			}
+			envVars = append(envVars, fmt.Sprintf("%s=%s", file.Name(), string(data)))
+		}
+	}
+	containerConfig := &container.Config{
+		Hostname: os.Hostname(),
+		User:     l.RunAs,
+		Env:      envVars,
+		Image:    l.Image,
+	}
+
+	restartPolicy := "always"
+	switch l.RestartPolicy_ {
+	case runit.RestartPolicyAlways:
+		// keep it the way it is
+	case runit.RestartPolicyNever:
+		restartPolicy = "no"
+	default:
+		return util.Errorf("invalid restart policy: %s", l.RestartPolicy_)
+	}
+
+	bindings := []string{}
+	dockerMountsDir := filepath.Join(l.RootDir, "docker_mounts.d")
+	mountFiles, err := ioutil.ReadDir(dockerMountsDir)
+	if err != nil && !os.IsNotExist(err) {
+		return util.Errorf("could not read from %s dir: %v", dockerMountsDir, err)
+	}
+	for _, file := range mountFiles {
+		mounts := Mounts{}
+		if !file.IsDir() {
+			// read content and load into bindings var
+			mountFilePath := filepath.Join(dockerMountsDir, file.Name())
+			data, err := ioutil.ReadFile(mountFilePath)
+			if err != nil {
+				return util.Errorf("could not read file %s: %v", mountFilePath, err)
+			}
+			yaml.Unmarshal(data, mounts)
+			for _, mount := range mounts {
+				if mount.Type != "bind" {
+					continue
+				}
+				bindings = append(bindings, fmt.Sprintf("%s:%s", mount.Source, mount.Target))
+			}
+		}
+	}
+	hostConfig := &container.HostConfig{
+		Binds:       bindings,
```

Comment:
I wonder if we want to use the `Mounts` field instead of `Binds`. From https://docs.docker.com/storage/bind-mounts/, it sounds like --mounts is newer and more explicit and generally suggested for new users.

It also sounds like Binds will create the directory in the container even if the source folder doesn't exist?

Reply:
From that link:
*"By contrast, when you use a volume, a new directory is created within Docker’s storage directory on the host machine, and Docker manages that directory’s contents."*
I don't think we want or need that.

*"It also sounds like Binds will create the directory in the container even if the source folder doesn't exist?"*
From that link: *"The file or directory does not need to exist on the Docker host already. It is created on demand if it does not yet exist. Bind mounts are very performant, but they rely on the host machine’s filesystem having a specific directory structure available."*
From my understanding, for binds, the source directory needs to exist. Even without docker the directories we want to bind need to exist for the app to run.